### PR TITLE
#1339 - Password Change Screen

### LINKED ISSUE #1339: [Add Feature] Change Password Option in the Edit Profile section
Part of the Feature Ratings from July. 

See below for wireframe of feature. Created a simple tab in the edit portfolio called `Security Information`. This is a similar format to change passwords as other large Social Feeds (Instagram, Snapchat, and Facebook) 

Requirements:
1. User must enter their old password correctly in order to meet the requirements to update the new password.
2. The password input will have to hide the characters inputted by representing ********* upon typing.
3. New Password and confirmation password must match.
4. Forgot your password Link will follow the same format as the Forgot your password link on login page. You will have to verify your email and a link to change password will be provided in that email. 

Password Reset in Edit Profile

### PR DESCRIPTION
**Pull request by Carrany**
### What does this pull request aim to achieve? 💯
Create a Change password screen
_Please be concise and link any related issue(s):_
resolves #1339

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

### CODE REVIEW

**Review by nadinejuraschek — DISMISSED** [review]
I'm not sure what the design looks like for this page, but I am sure the button isn't styled correctly: 
Screen Shot 2020-10-21 at 1 33 46 PM
It should match the other buttons on our page (purple, no blue border, white text, no box-shadow), could you re-use a button component we already have?

**Review by galeza — DISMISSED** [review]
Do we have any design in Figma or Zeplin for this feature?

I think that the design, components are completely different from the style that we have on (for example) Join In screen. Example:
Screen Shot 2020-10-22 at 11 59 52

**Review by rabbi08 — CHANGES_REQUESTED** [review]
Following issues have been observed while testing:
- The 'New Password' textbox stretch as soon as user start typing.
- The password instructions should not be displayed in single line, as that is the reason it the textbox length vary. the password instruction can be displayed in bulleted list.

ezgif com-gif-maker (25)

- The term 'New Password, confirmation' is not user friendly, it should be 'Confirm Password'

Functional Issue: 
1. After entering all valid values, the password change is unsuccessful. Refer attached screenshot.
2. In the Password change message we have 2 different 'x' close icons which doesn't look appropriate.

image

Do we have defined zeplin/figma design to follow.

**Review by rabbi08 — CHANGES_REQUESTED** [review]
@Carrany 
Is the backend of this features ready to test i.e. Password change functionality is implemented or currently we have to test the UI screen of password change feature. Thanks.

As currently, still after clicking on 'Save Changes' -> The password change is unsuccessful and all the fields are reset. Please see below screenshot,

image

UI Issue:
The 'Edit Profile' option screen, have white screen on the top and all the options are not displayed properly, after scrolling down , there is white space is at the bottom too. Can we adjust all the 3 options of EDIT Profile, without scroll bar by removing the white space?
Please see below GIF.

ezgif com-gif-maker (27)

Also, please confirm on the following missing text and link as per the design mentioned in #1339 
- The page should have title as 'Change Your Password'
- Link to 'Forgot Password' should be there on the page.

**Comment by Carrany** [comment]
@rabbi08 
Backend is ready to test and is working accordingly. You could check if the current password is the correct one

**Comment by rabbi08** [comment]
@Carrany 
Thanks for confirming. 
I am able to change my password successfully. Please see below GIF.
ezgif com-gif-maker (28)

I have following observation, kindly look into these too:
- The 'x' and 'checkbox' sign in the 'Password change successfully/unsuccessfully' should not be there, the message is enough with red cross at the end of the line.
- There are no validation to guide the user if the password change is unsuccessful as if:
  a. if user enter wrong current password, No error message to guide the user.

**Review by joshmorel — CHANGES_REQUESTED** [review]
We shouldn't commit the `yarn.lock` files. Since we are using `package-lock.json` I think it is best to use npm for installing project dependencies (handled automatically be docker). But in any case, we can add .gitignore for these (I will do). 

**This has been fixed**

**Review by joshmorel — CHANGES_REQUESTED** [review]
@Carrany as per @rabbi08 question about the wrong existing password. Can we use something similar to Login.js . It is translating the error message from the backend (already translated and it will be the same as with wrong credentials for login).

https://github.com/FightPandemics/FightPandemics/blob/b347a0392498cb682c1de4e1d1110903b0020d62/client/src/pages/Login.js#L267

**Review by rabbi08 — DISMISSED** [review]
Unable to open EDIT options for individual profile, as after user clicks on Edit icon the screen turn grey on mobile. But the edit icon works fine for org profile. 

Please see below gif
image

File: client/src/components/EditProfile/EditComponents.js [hunk 1]
```diff
@@ -205,3 +205,26 @@ export const PlusIcon = styled.img`
   transform: translate(-50%, -50%);
   position: absolute;
 `;
+export const CustomContainer = styled.div`
```
Comment (joshmorel) [inline client/src/components/EditProfile/EditComponents.js]: I don't see this ones used anywhere.

**Review by joshmorel — CHANGES_REQUESTED** [review]
Thanks @Carrany . So long as the other issues are confirmed resolved I'm happy with one minor things. Those unused styled components we should remove (I guess you started out using them then switched back to the existing ones used by EditAccount etc?).

Another thing I think we should resolve at a design level. We should have a consistent approach to redirect after editing user/org. In most we are redirecting to profile page but not `edit-organisation-account/` and not here. Actually re-directing back to profile is non-standard UX so it might be better to stay in place BUT we need a better way to navigate on mobile.

**Comment by rabbi08** [comment]
@Carrany 
- The page title should be 'Change Your Password' not 'Account Information'

Though below scenarios are working fine:
- I am able to change my password- Pass
- wrong current pwd err message is displayed

**Comment by Carrany** [comment]
@rabbi08 

Account Information is a link to editing account information, not the title of the page. The title of the page is Edit Profile

**Comment by rabbi08** [comment]
@Carrany 
On Desktop, the page title is 'Edit Profile' and the TABs have respective names.

But on mobile, every page has its own name: 
- Account Information
- Edit profile 
- "Account Information" for Password change too. Please see below GIF.

ezgif com-video-to-gif 7

**Comment by Carrany** [comment]
@rabbi08 Please confirm the change from Account Information to Change password on mobile view

**Review by rabbi08 — APPROVED** [review]
Thanks @Carrany 

Verified on mobile and desktop on feature branch as working fine:
- user is able update the pwd successfully
- valid err message is displayed if user enter wrong current pwd

Please see below GIF
image

**Review by galeza — CHANGES_REQUESTED** [review]
@Carrany 

I have a question about a button: Save Changes. 
On the Join Now screen when passwords don't match we keep button Join Now disabled
with an attribute aria-disabled="true".
Screen Shot 2020-11-06 at 15 01 02

On the change password edit-security screen Save Changes button looks like enabled - with an attribute aria-disabled="false". 
Maybe it should be : aria-disabled="true" **when passwords don't match**.
Screen Shot 2020-11-06 at 14 54 08

**Comment by rabbi08** [comment]
@Carrany can we fix this PR and get it merged by today i.e. Monday, Nov 09 EOD, so that we can work on GTM tags for this feature and release to PROD by Tuesday/Wed (Nov 10). As this feature needs GTM tag assigned and translation done in STG environment. Thanks.

**Review by nadinejuraschek — APPROVED** [review]

**Comment by rabbi08** [comment]
Verified the fix on feature branch, the 'Save changes' button remain disable until user enters all the details. 
Thanks

**Review by Zaydme — APPROVED** [review]
LGTM

**Review by galeza — APPROVED** [review]
LGTM

File: backend/lib/endpoints/auth.js [hunk 2]
```diff
@@ -159,11 +169,57 @@ async function routes(app) {
         );
         throw app.httpErrors.tooManyRequests("maxSignInAttemptsExceeded");
       }
-      req.log.error(err, "Error logging in");
```
Comment (ddveloper) [inline backend/lib/endpoints/auth.js:162]: Will there be other error cases need these 2 lines?

**Pull request merged**